### DISCUSSION TITLE
CPS, ASR, ACD ...

### DISCUSSION THREAD

**Discussion by vadimd333** [sipcapture/homer#566] [Q&A] [answered]
Hello!

How can I get CPS, ASR, ACD in Homer 7?

**Comment by lmangani** [comment]
TLDR; doing math on the METHODS and their statistics. Some [examples](https://grafana.com/grafana/dashboards/11765-sip-kpi-s/) are available in the prometheus Grafana dashboards

<img src="https://grafana.com/api/dashboards/11765/images/7601/image">

**Comment by vadimd333** [comment]
Thanks! 

How can I get a KPI from a specific IP-address? 

The query is: `sum without(direction,node_id) (rate(heplify_method_response{target_name=~"$tn", method="INVITE"}[10m]))`

What do i need to add to this query?

**Comment by adubovikov** [comment, marked as answer]
Hi @vadimd333,

Following up on the KPI-by-IP question.

Homer 7 does **not** show CPS / ASR / ACD inside the Homer UI itself. You get them from **heplify-server Prometheus metrics + Grafana** (as [@lmangani](https://github.com/lmangani) pointed out). See the [SIP KPI Grafana dashboard](https://grafana.com/grafana/dashboards/11765-sip-kpi-s/) for ready-made panels.

### Metric source

heplify-server exposes (among others):

- `heplify_method_response{target_name, direction, node_id, response, method}`
- `heplify_kpi_srd` (INVITE setup delay)
- RTCP / XRTP metrics for quality

Homer 11 uses the same idea with `homer_sip_method_response_total` and related counters.

### CPS / ASR / ACD (typical PromQL)

**CPS** (INVITE attempts per second):

```promql
sum without(direction, node_id) (
  rate(heplify_method_response{target_name=~"$tn", method="INVITE"}[5m])
)
```

**ASR** (example: 200 OK on INVITE / all INVITE attempts):

```promql
sum without(direction, node_id) (
  rate(heplify_method_response{target_name=~"$tn", method="INVITE", response="200"}[5m])
)
/
sum without(direction, node_id) (
  rate(heplify_method_response{target_name=~"$tn", method="INVITE"}[5m])
)
```

**ACD** depends on how you define "answered call duration". Often computed from CDR/BYE timing or custom panels in the SIP KPI dashboard, not a single built-in Homer UI field.

Adjust the `[5m]` / `[10m]` window to your scrape interval.

### Filter by a specific IP address

Important: `heplify_method_response` has **no `src_ip` / `dst_ip` label**.

It uses **`target_name`**, mapped from packet IPs in **heplify-server** config:

```
PromTargetIP=10.1.2.3,10.1.2.4
PromTargetName=trunk_a,trunk_b
```

When traffic matches `PromTargetIP`, metrics get that `target_name`. `direction` is `src` or `dst`.

So to KPI one IP:

1. Add the IP to `PromTargetIP` and give it a name in `PromTargetName` (can be the IP itself, e.g. `10.1.2.3`).
2. Restart/reload heplify-server with Prometheus export enabled.
3. Filter in Grafana:

```promql
sum without(direction, node_id) (
  rate(heplify_method_response{target_name="10.1.2.3", method="INVITE"}[10m])
)
```

Or only traffic where that host was **source**:

```promql
rate(heplify_method_response{target_name="10.1.2.3", direction="src", method="INVITE"}[10m])
```

Check available labels in Prometheus / Grafana Explore:

```promql
group by (target_name, direction) (heplify_method_response)
```

### Requirements

- heplify-server: `-prometheus` / `Prometheus = true` in config
- Prometheus scraping heplify metrics endpoint
- Grafana with the SIP KPI dashboard (or your own panels)

If you share your heplify `PromTargetIP` / `PromTargetName` setup and which IP you want, we can write the exact `$tn` variable query for your dashboard.
